File: src/engines/mbti/services/elementMapper.py
"""
Converts normalized cognitive function scores into a 5-element vector.

Uses the element matrix from parameters.py:
  For each function, multiply its normalized score by that function's
  element row, then sum across all 8 functions.

Finally, normalize the 5 elements so they sum to 1.0.
"""
from ..constants import parameters

ELEMENTS = ["air", "water", "fire", "earth", "spirit"]


def functions_to_elements(normalized_scores):
    """
    Takes normalized function scores (sum to 1.0) and returns
    a normalized 5-element vector (sum to 1.0).

    Steps:
      1. For each function, multiply its weight by its element row.
      2. Sum all contributions per element.
      3. Normalize so elements sum to 1.0.
    """
    raw_elements = {el: 0.0 for el in ELEMENTS}

    for fn, weight in normalized_scores.items():
        row = parameters.ELEMENT_MATRIX[fn]
        for el in ELEMENTS:
            raw_elements[el] += weight * row[el]

    # Normalize to sum to 1.0
    total = sum(raw_elements.values())
    if total == 0:
        return {el: 0.2 for el in ELEMENTS}

    return {el: round(raw_elements[el] / total, 4) for el in raw_elements}

File: src/server/engines/mbti/elementMapper.ts
import type { ElementVector } from "@/types/personalization";
import { ELEMENT_MATRIX, FUNCTIONS } from "./constants";

const ELEMENTS = ["air", "water", "fire", "earth", "spirit"] as const;

/**
 * Convert normalized cognitive function scores into a 5-element vector.
 *
 * For each function, multiply its weight by its ELEMENT_MATRIX row,
 * then accumulate per element and normalize to sum = 1.0.
 * Returns 0.2 each if all element values are zero.
 */
export function functionsToElements(
  normalized: Record<string, number>,
): ElementVector {
  const raw: Record<string, number> = {
    air: 0, water: 0, fire: 0, earth: 0, spirit: 0,
  };

  for (const fn of FUNCTIONS) {
    const weight = normalized[fn] ?? 0;
    const row = ELEMENT_MATRIX[fn];
    for (const el of ELEMENTS) {
      raw[el] += weight * row[el];
    }
  }

  const total = ELEMENTS.reduce((sum, el) => sum + raw[el], 0);

  if (total === 0) {
    return { air: 0.2, water: 0.2, fire: 0.2, earth: 0.2, spirit: 0.2 };
  }

  return {
    air: Math.round((raw.air / total) * 1e4) / 1e4,
    water: Math.round((raw.water / total) * 1e4) / 1e4,
    fire: Math.round((raw.fire / total) * 1e4) / 1e4,
    earth: Math.round((raw.earth / total) * 1e4) / 1e4,
    spirit: Math.round((raw.spirit / total) * 1e4) / 1e4,
  };
}
